### ISSUE TITLE
[bitnami/redis-cluster] Graceful restarts 

### ISSUE THREAD

**Issue by shettyvaishak** [bitnami/charts#13583]
### Name and Version

bitnami/redis-cluster 7.5.6

### What steps will reproduce the bug?

Keep some clients connected to the cluster 

Perform a rolling restart of the redis statefulset 
kubectl rollout restart statefulset <statefulset> -n <namespace>

### What is the expected behavior?

Client connections should be gracefully handled before deleting the pods

### What do you see instead?

Client connections are not handled and the pods are terminated causing client to throw errors

### Additional information

We have a scale test environment where we perform tests on the cluster with high number of users. We have noticed that whenever we restart redis pods there is a spike in number of 5XX errors, we believe it is because the connections are not being gracefully handled before terminating the pod. 

Is there any best practice to handle these connections before terminating the pod by running a script through a pre stop lifecycle hook??

**Comment by javsalgar** [comment]
Hi,

For this kind of scenarios, we would normally use lifecycle hooks to ensure a proper termination of the solution. In the case of redis, as you have experience with this workload, do you recommend any specific graceful shutdown command that we could add to the redis statefulset?

**Comment by shettyvaishak** [comment]
Tried running the CLIENT list command in the pre-stop hook to see if waiting for clients to close when the pod is put into the terminating state would reduce the 5xx , but it did not help . The client list doesnt seem to reduce when pod is put in terminating state . This is the script we tested in the hook 

```
          end=$((SECONDS+30))
          client_closed=false
          while [[ $SECONDS -lt $end ]]; do
              client_list=$(redis-cli client list type normal)
              count=0
              client_count=0
              echo -e "$client_list \n" > /proc/1/fd/1
              for client in $client_list; do
                  if [[ $(( count % 23 )) == 0 ]]; then
                      client_count=$((client_count+1))
                  fi
                  count=$((count+1))
              done
              if [[ $client_count -le 1 ]]; then
                  client_closed=true
                  break
              else
                  sleep 5s
              fi
          done
          if $client_closed; then
              echo "All client connections closed proceeding with Redis shutdown" > /proc/1/fd/1
          else
              echo "Shutting down redis before closing all client connections. Number of clients connected: $client_count" > /proc/1/fd/1
          fi
```
Noticed that most of the 5xx are when pod which is redis master is being restarted. Tried doing a manual failover for master to see if there are any 5xx's but there was none. So next thinking about doing a failover for master in the prestop hook if the pod being restarted is a master.

Let me know if there are any other suggestions which we could try out?

**Comment by javsalgar** [comment]
Hi,

I'm not an expert but maybe the prestop hook in the master as you were mentioning helps with the issue.

**Comment by shettyvaishak** [comment]
For us this issue was mostly occuring when a pod which was redis master was restarting. I did a manual failover to see if it is causing the 5xx and it didnt. When redis recieves a sigterm it initiates the shutdown process and this pauses all the clients. Since our clients mostly connects to master it was causing issues when the pods were restarting. I have been able to resolve this issue by doing a failover in the preStop hook so that every time only the slave is getting shutdown
